### PR TITLE
IN: Messaging #87

### PR DESCRIPTION
**Pull request by jstyle5**
### What this PR does (required):
- Issue Link: https://github.com/hatchways/team-breadsticks/issues/87
- Created `conversation context`
- Refactored `seed` related files.

#### Desktop version (Assume the logged-in user is `DemoUser`, and he already has 3 conversations)
- The latest message for `DemoUser` is `DemoUser ---> Test3`. So, it is placed at the top of the `ChatSideBanner`.
>
Screen Shot 2021-10-05 at 17 24 03
---

#### Desktop version (After clicking the conversation with `Test3` (ActiveConvo: Test3))
Screen Shot 2021-10-05 at 17 29 57
---

#### Desktop version (After sending a message to `Test3`, the latestMessage is also updated)
Screen Shot 2021-10-05 at 17 30 45
---

#### Desktop version (After clicking the conversation with `Test1` (ActiveConvo: Test1))
Screen Shot 2021-10-05 at 17 32 32
---

#### Desktop version (After sending a message to `Test1`, the latestMessage is also updated. And, the `ChatBox(Conversation with Test1`) is placed at the top of the `ChatSideBanner`.
Screen Shot 2021-10-05 at 17 36 44
---

#### Desktop version (Open a new browser and log-in as `Test1`, you can see the message from `DemoUser` that was just sent before)
Screen Shot 2021-10-05 at 17 39 15
---

#### Mobile version (Assume the logged-in user is `DemoUser`, and he already has 3 conversations)
Screen Shot 2021-10-05 at 17 42 33
---

#### Mobile version (After clicking the conversation with `Test2 (ActiveConvo: Test2)`
Screen Shot 2021-10-05 at 17 43 42
---

#### Mobile version (After sending a message to `Test2`)
Screen Shot 2021-10-05 at 17 44 25
---

#### Mobile version (If you want to go back to `ChatSideBanner`, swipe the screen from the left to the right)
Screen Shot 2021-10-05 at 17 45 42
---

#### Mobile version (After sending a message to `Test2`, the latestMessage is also updated. And, the `ChatBox(Conversation with Test2`) is placed at the top of the `ChatSideBanner`.
Screen Shot 2021-10-05 at 17 48 18

#### Mobile version (Open a new browser and log-in as `Test2`, you can see the message from `DemoUser` that was just sent before)
Screen Shot 2021-10-05 at 17 51 33
---

### Any information needed to test this feature (required):
- In order to create `Conversation`, **users must have their own `Profiles`.**
- Go into the server directory `cd server`
- Create `.env`, or if it exists, create your environment variables. `JWT_SECRET="Your JWT SECRET"
MONGO_URI="Your MongoDB URI"`
- Run `npm run seed` to import mock-data (NOTE: Whenever this command is executed, it will first remove all existing documents in `User`, `Profile`, `Conversation`, and `Message` collections. Then, it will insert mock-data automatically)
- Run `npm run dev` and test routes with Postman
- Since these routes are private, you need to log in first.

### Any issues with the current functionality (optional):
- `Search` feature is not implemented with this version. So, users can exchange messages within existing conversations only for now.
- Since `socket.io` is not set up with this current version, the real-time conversation feature is not considered.
- `Online status` and `Number of unread messages` are not considered.

### CODE REVIEW

**Comment by ezeilo-su** [comment]
There's a conflict to be resolved in this PR, @jstyles5

File: server/seed/seed.js [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+async function deleteData() {
+  try {
+    await User.deleteMany();
+    console.log("Users are deleted..");
+
+    await Conversation.deleteMany();
+    console.log("Conversations are deleted..");
+
+    await Message.deleteMany();
+    console.log("Messages are deleted..");
```
Comment (ezeilo-su) [inline server/seed/seed.js:22]: Remove this console logs

File: server/seed/seed.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    await Profile.deleteMany();
+    console.log("Profiles are deleted..\n");
+  } catch (err) {
+    console.error(err);
```
Comment (ezeilo-su) [inline server/seed/seed.js:27]: Remove this console logs

File: server/seed/seed.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+  }
+}
+
+async function seed() {
+  await connectDB();
+  await deleteData();
+
+  // Total 9 users.
+  // DemoUser: testUsers[0], OtherUsers: testUsers[1] ~ testUsers[8]
```
Comment (ezeilo-su) [inline server/seed/seed.js:36]: Remove comments

File: server/seed/seed.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  const testUsers = await seedUsers();
+  console.log(testUsers.length + " testUsers are created..");
+
+  // Total 9 profiles.
+  // DemoUserProfile: testProfiles[0], OtherUserProfiles: testProfiles[1] ~ testProfiles[8]
```
Comment (ezeilo-su) [inline server/seed/seed.js:41]: Remove comments

File: server/seed/seed.js [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  const testProfiles = await seedProfiles(testUsers);
+  console.log(testProfiles.length + " testProfiles are created..");
+
+  // Initial conversation between demoUser and test1
```
Comment (ezeilo-su) [inline server/seed/seed.js:45]: Remove comments

File: server/seed/seed.js [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+async function runSeed() {
+  console.log("seeding..");
+  try {
+    await seed();
+    console.log("data seeded successfully..");
+  } catch (err) {
+    console.error(err);
+    process.exitCode = 1;
+  } finally {
+    console.log("MongoDB connection closed..".cyan.underline + "\n");
```
Comment (ezeilo-su) [inline server/seed/seed.js:145]: No console log is needed

**Review by ezeilo-su — CHANGES_REQUESTED** [review]
Well done on this one :+1: 
Just implement the requested changes before approval.
Happy coding!

File: client/src/App.tsx [hunk 7]
```diff
@@ -24,25 +24,25 @@ function App(): JSX.Element {
...
-                  <Route exact path="/messages" component={Messages} />
-                </ConvoProvider>
-                <Route exact path="/my-sitters" component={MySitters} />
-                <Route exact path="/my-jobs" component={MyJobs} />
-                <Route path="*">
-                  <Redirect to="/login" />
-                </Route>
-              </Switch>
-            </SocketProvider>
+            {/* <SocketProvider> */}
```
[the remark this answers is not in the document]
Comment (jstyle5) [reply client/src/App.tsx:27]: Restored

File: server/seed/seed.js (hunk 3 above)
Comment (jstyle5) [reply server/seed/seed.js:36]: Applied

File: server/seed/seed.js (hunk 4 above)
Comment (jstyle5) [reply server/seed/seed.js:41]: Applied

File: server/seed/seed.js [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+  // Total 9 users.
+  // DemoUser: testUsers[0], OtherUsers: testUsers[1] ~ testUsers[8]
+  const testUsers = await seedUsers();
+  console.log(testUsers.length + " testUsers are created..");
+
+  // Total 9 profiles.
+  // DemoUserProfile: testProfiles[0], OtherUserProfiles: testProfiles[1] ~ testProfiles[8]
+  const testProfiles = await seedProfiles(testUsers);
+  console.log(testProfiles.length + " testProfiles are created..");
```
[the remark this answers is not in the document]
Comment (jstyle5) [reply server/seed/seed.js:43]: Applied

File: server/seed/seed.js [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    conversation: demoUserConvo._id,
+    senderProfile: testProfiles[1]._id,
+    text: "Test1 ---> DemoUser",
+    recipientRead: false,
+  });
+
+  demoUserConvo.latestMessage = demoUserTest1Message3._id;
+  await demoUserConvo.save();
+
+  // Initial conversation between demoUser and test2
```
[the remark this answers is not in the document]
Comment (jstyle5) [reply server/seed/seed.js:75]: Applied

**Comment by jstyle5** [comment]
> Well done on this one 👍 Just implement the requested changes and resolve merge conflict before approval. Happy coding!

Thank you :)

File: client/src/context/useConvoContext.tsx [hunk 10]
```diff
@@ -52,21 +63,47 @@ export const ConvoProvider: FunctionComponent = ({ children }): JSX.Element => {
...
+      }
     },
     [activateChat],
   );
 
+  const sendMessageContenxt = useCallback(
+    async (sendingMessageInfo: Message) => {
+      const sendingMessage = await sendMessage(sendingMessageInfo);
+      setActiveConvo(
+        produce((draft) => {
```
Comment (rajivtitus) [inline client/src/context/useConvoContext.tsx:81]: 👍🏽

**Review by rajivtitus — APPROVED** [review]
Great work with this! 💯 

Just a minor optinal comment 😄

File: server/controllers/conversation.js [hunk 11]
```diff
@@ -9,14 +9,42 @@ exports.getAllConversations = asyncHandler(async (req, res, next) => {
...
+      $or: [
+        { user1Profile: userProfile._id },
+        { user2Profile: userProfile._id },
+      ],
     })
       .populate("latestMessage")
-      .populate("user1", "-password -register_date -__v")
-      .populate("user2", "-password -register_date -__v");
+      .populate("user1Profile", "userId photos firstName lastName")
+      .populate("user2Profile", "userId photos firstName lastName");
```
[the remark this answers is not in the document]
Comment (jstyle5) [reply server/controllers/conversation.js:27]: @rajivtitus I just tested it with `.populate(...).sort({'lastMessage.createdAt': -1}), but it seems that it's not working in the current `mongoose`? 🤔  (https://github.com/Automattic/mongoose/issues/4481)

[the remark this answers is not in the document]
Comment (jstyle5) [reply server/controllers/conversation.js:27]: @rajivtitus Many people do want `sort` feature after `populate`. It's very sad that is not supported. 😭

**Pull request merged**